### ISSUE TITLE
Performance improvements

### ISSUE THREAD

**Issue by ndorfin** [nzruby/conf#30]
This isn't so much an issue as a log of performance improvements to the site.

# Test setup

## [WebPageTest](https://www.webpagetest.org/)

Node: Australia, EC2
Browser: Chrome
Connection: Cable
Number of passes: 3
Type: First and repeat view

## Lighthouse 2.0
Running Chrome's new Audit feature: Lighthouse.

# WebPageTests
Date | SpeedIndex | Description 
------|-----------|--------------
[2017-08-24 11PM](https://www.webpagetest.org/result/170824_NA_6107a394107fd98c95933811e2478758/) | 3900 | Baseline
[2017-09-06 9PM](https://www.webpagetest.org/result/170906_MX_134b2078407fd429bad17d360910f260/) | 3000 ✅  | Images reformatted: 1.5MB to 0.9MB
[2017-09-09 1PM](https://www.webpagetest.org/result/170909_N0_496fd42ded0edf639dc56477e62a1950/) | 2635 ✅  | HTTPS, CloudFlare enabled
[2017-09-10 11PM](https://www.webpagetest.org/result/170910_52_1be4fb88daa4eba72ff355d2e4e7a5da/) | 2660 🔺 | Less Fonts, Static Map, Slow jQ CDN
[2017-09-16 12PM](https://www.webpagetest.org/result/170915_8Y_85fdc0be485d90395eb994d1cd188317/) | 2372 ✅ | No jQuery, Asset hashing, 550KB
[2017-09-26 5PM](https://www.webpagetest.org/result/170926_BC_c453ef5be568254bb1c0f64aa967c2f9/) | 2212 ✅ | Preload, async, turbolinks, less fonts
[2017-09-28 8AM](https://www.webpagetest.org/result/170927_NT_32edb18adbcfe1742df49567eab33fec/) | 2008 ✅ | ServiceWorker, Inline CSS 
[2017-09-29 10PM](https://www.webpagetest.org/result/170929_NB_281a7ad4dd0a20ee8b28533b84ee5832/) | 1908 ✅ | Clean up some of the SW pre-cache
[2017-11-01 10AM](https://www.webpagetest.org/result/171031_E2_0d3df1f11c6dda27eed8e8a1af828d64/) | 599 ✅  | Self-hosted fonts, CSS cleanup

# Lighthouse Tests
Date | PWA | Perf | A11y | Practices | Description
-----|------|--------------|--------------|-----------|------------
Aug-24 | 27 | 34 | 94 | 58 | Baseline
Sep-06 | 27 | 50 ✅  | 94 | 58 | Images reformatted
Sep-09| 45 ✅  | 54 ✅  | 94 | 75 ✅  | CloudFlare, HTTPS, PWA
Sep-10 | 45 | 72 ✅  | 97 ✅  | 83 ✅ | Less fonts, Static map
Sep-16 | 55 ✅ | 88 ✅ | 97 | 85 ✅ | No jQuery, Asset hashing 
Sep-18 | 73 ✅ | 83 🔻 | 97 | 100 ⭐️ | Manifest and link fixes
Sep-26 | 73 ✅ | 91 ✅  | 97 | 100 ⭐️ | Preload, async, turbolinks
Sep-28 | 100 ⭐️ | 97 ✅  | 97 | 100 ⭐️ | ServiceWorker, Inline CSS
Nov-01| 100 ⭐️ | 97  | 97 | 100 ⭐️ | Self-hosted fonts, CSS cleanup

**Comment by ndorfin** [comment]
The first set of [webpagetest comparisons are now visible](https://www.webpagetest.org/video/compare.php?tests=170906_MX_134b2078407fd429bad17d360910f260,170824_NA_6107a394107fd98c95933811e2478758).

Removing unnecessary data from the images and resizing them has reduced the total payload from 1.5MB to 0.9MB for the homepage. The tests had a marginal increase in TTFB (Time to first byte) which would have had a knock-on effect for other requests, resulting in a first render reduction from 3.9s to 3s.

Lazy-loading the fonts, inlining critical CSS, and [reducing dependencies on scripts](https://github.com/nzruby/conf/pull/33) will pull us closer to the magical 1s render mark.

**Comment by ndorfin** [comment]
With CloudFlare and HTTPS enabled and the beginnings of a PWA put in place, both the WebPageTest and Lighthouse scores improved.

The time to first paint has dropped from 3.0s to 2.1s, mostly helped along by CloudFlare's infrastructure, probably because there's an edge node in Australia.

Here's the [latest WebPageTest comparison](https://www.webpagetest.org/video/compare.php?tests=170909_N0_496fd42ded0edf639dc56477e62a1950%2C170906_MX_134b2078407fd429bad17d360910f260%2C170824_NA_6107a394107fd98c95933811e2478758&thumbSize=200&ival=100&end=visual)

**Comment by mermop** [comment]
How does it look after #44 #42 and #40 @ndorfin ?

**Comment by ndorfin** [comment]
@mermop I'm waiting for the CloudFlare cache to be invalidated.

http://kiwi.ruby.nz (Github) is looking good,
but https://kiwi.ruby.nz (CloudFlare) is one release behind.

It might be worth adding an invalidation step to after `middleman-deploy` has done its job.

Also, can we force HTTP --> HTTPS redirects? CloudFlare has a setting for that.
cc @steveh

**Comment by mermop** [comment]
Do you want to put some comments in #45 about that - would fingerprinting assets sort out Cloudflare or is there an additional step we'd need to take?

**Comment by ndorfin** [comment]
We're seeing some [interesting changes in the WebPageTest comparisons now](https://www.webpagetest.org/video/compare.php?tests=170910_52_1be4fb88daa4eba72ff355d2e4e7a5da%2C170909_N0_496fd42ded0edf639dc56477e62a1950%2C170906_MX_134b2078407fd429bad17d360910f260%2C170824_NA_6107a394107fd98c95933811e2478758&thumbSize=200&ival=100&end=visual).

While the TTFB is the same, the impact of the jQuery CDN being slower today has caused the conf site to do its first paint slower. The text shows up earlier though, so that's great!
We're also down to 579KB! Almost a third of what it originally was!

Render blocking scripts should come last, or at least be deferred / async. Using the jQuery package from a third-party CDN incurs extra SSL handshakes over HTTPS, but if we refactor the main nav behaviour on mobile, we should be able to remove jQuery (and it's third-party CDN) as a dependency. I've issued #49 to accommodate this.

**Comment by ndorfin** [comment]
After removing jQuery, we're seeing:

- less time spent rendering,
- painting is happening quicker
- less CPU usage
- JS down to less than 1KB

You can [see the comparison on WebPageTest](https://www.webpagetest.org/video/compare.php?tests=170915_8Y_85fdc0be485d90395eb994d1cd188317,170913_3F_57af1d5861cc8ca8bb0f55d5aac29d21,170909_N0_496fd42ded0edf639dc56477e62a1950,170906_MX_134b2078407fd429bad17d360910f260,170824_NA_6107a394107fd98c95933811e2478758)

**Comment by mermop** [comment]
What's the difference between `no-jquery` and `no-jquery-7` there? There are a couple of things that are quite different between them -

<img width="889" alt="screen shot 2017-09-17 at 3 45 51 pm" src="https://user-images.githubusercontent.com/1615322/30517826-64bd5058-9bbf-11e7-9d6b-7f600496eacd.png">

**Comment by ndorfin** [comment]
@mermop The main difference is the Time To First Byte.
I ran several tests since the last release, but most of the older tests were running a bit slow between EC2 in Sydney and CloudFlare's end in Auckland. The test labelled `no-jquery-ttfb` has a similar TTFB as the `https` and `-fonts -map` tests so they make for better comparisons.

<img width="851" alt="screen shot 2017-09-17 at 4 38 34 pm" src="https://user-images.githubusercontent.com/107569/30518073-560933bc-9bc7-11e7-8ca9-9064ce0657e5.png">

**Comment by mermop** [comment]
Seeing this on lighthouse now - I think largely from removing that second font in #63 
<img width="561" alt="screen shot 2017-09-26 at 4 45 08 pm" src="https://user-images.githubusercontent.com/1615322/30841669-174b8ea8-a2da-11e7-9800-952bf517b148.png">

**Comment by ndorfin** [comment]
The [comparisons are looking even better](https://www.webpagetest.org/video/compare.php?tests=170926_BC_c453ef5be568254bb1c0f64aa967c2f9,170915_8Y_85fdc0be485d90395eb994d1cd188317,170913_3F_57af1d5861cc8ca8bb0f55d5aac29d21,170909_N0_496fd42ded0edf639dc56477e62a1950,170906_MX_134b2078407fd429bad17d360910f260,170824_NA_6107a394107fd98c95933811e2478758) now @mermop - good work!

**Comment by mermop** [comment]
Down even further after #59 😎

**Comment by mermop** [comment]
Look at this!
<img width="561" alt="screen shot 2017-09-27 at 11 59 24 pm" src="https://user-images.githubusercontent.com/1615322/30910108-f9c7483e-a3df-11e7-8e32-2801442508e4.png">

Nice work @ndorfin 🆒 😎

**Comment by ndorfin** [comment]
@mermop @steveh Whoop-whoooooop!! 🎆  Thanks for your patience :)

I ran a test this morning, and even with a disastrous TTFB, it [rendered faster than all previous tests by 300ms](https://www.webpagetest.org/video/compare.php?tests=170927_NT_32edb18adbcfe1742df49567eab33fec,170926_FW_4d87e7a56134c9b1566cafd220c3ca99,170915_8Y_85fdc0be485d90395eb994d1cd188317,170913_3F_57af1d5861cc8ca8bb0f55d5aac29d21,170909_N0_496fd42ded0edf639dc56477e62a1950,170906_MX_134b2078407fd429bad17d360910f260,170824_NA_6107a394107fd98c95933811e2478758)!

It looks like we can clean up the pre-cache a little, I'll issue a PR for that this evening.

**Comment by ndorfin** [comment]
We've still got a lot of unused files in the SW cache, so I've issued #73 to address that.
The [latest comparison](https://www.webpagetest.org/video/compare.php?tests=170929_NB_281a7ad4dd0a20ee8b28533b84ee5832,170927_46_f1a4367e5104e1596940845b8ecbe8c9,170926_FW_4d87e7a56134c9b1566cafd220c3ca99,170915_8Y_85fdc0be485d90395eb994d1cd188317,170913_3F_57af1d5861cc8ca8bb0f55d5aac29d21,170909_N0_496fd42ded0edf639dc56477e62a1950,170906_MX_134b2078407fd429bad17d360910f260,170824_NA_6107a394107fd98c95933811e2478758) was captured at what was a speedy response time at CloudFlare, with the rendering time down to 1.5s for first render.

We can do better by improving the webfont performance, so I created #76 to self-host the webfonts.

There's not much more we can do to improve the performance, save for switching CDN providers, trying longer cache times in CloudFlare, or rewriting the CSS - but it's been heaps of fun observing the progress. 💯

**Comment by ndorfin** [comment]
Self-hosting the webfonts has meant another 200ms off the time to render, bringing the SpeedIndex down to 599!
It looks like the Sydney EC2 testing instance has gone through a change recently, as the TTFB have halved, and the throughput seems a lot quicker. Regardless, the [comparison graphs are looking amazing](https://www.webpagetest.org/video/compare.php?tests=171031_E2_0d3df1f11c6dda27eed8e8a1af828d64,171004_SD_c6b57b9cc2f02dda1ca6cdf88859f9c9,170929_NB_281a7ad4dd0a20ee8b28533b84ee5832,170927_46_f1a4367e5104e1596940845b8ecbe8c9,170926_FW_4d87e7a56134c9b1566cafd220c3ca99,170915_8Y_85fdc0be485d90395eb994d1cd188317,170913_3F_57af1d5861cc8ca8bb0f55d5aac29d21,170909_N0_496fd42ded0edf639dc56477e62a1950,170906_MX_134b2078407fd429bad17d360910f260,170824_NA_6107a394107fd98c95933811e2478758)!

Good luck with the conference @mermop!
